### PR TITLE
feat(api): random

### PR DESCRIPTION
**Pull request by dcvz**

### CODE REVIEW

File: src/random.rs [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+            key.as_ptr(),
+            2,
+            stream.as_ptr(),
+        ))
+    };
+
+    (keys.index(0), keys.index(1))
+}
+
+/// Generate uniformly distributed random numbers with a given interval (`lower_bound:` and `upper_bound:`).
```
Comment (minghuaw) [inline src/random.rs:62]: I think we could borrow the python doc here, mention that the interval is half open, `[lower_bound, upper_bound)`

File: src/random.rs [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+/// let array = mlx_rs::random::uniform::<_, f32>(0, 10, &[50], &key);
+///
+/// // same, but in range `0.5 ..< 1`
+/// let array = mlx_rs::random::uniform::<_, f32>(0.5f32, 1f32, &[50], &key);
+/// ```
+#[default_device]
+pub fn uniform_device<'a, E: Into<Array>, T: ArrayElement>(
+    lower_bound: E,
+    upper_bound: E,
+    shape: &[i32],
```
Comment (minghuaw) [inline src/random.rs:80]: Is the shape optional?

File: src/random.rs [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+/// ```
+///
+/// # Params
+///  - shape: shape of the output, if `None` a single value is returned
+///  - loc: mean of the distribution, default is `0.0`
+///  - scale: standard deviation of the distribution, default is `1.0`
+///  - key: PRNG key
+#[default_device]
+pub fn normal_device<'a, T: ArrayElement>(
+    shape: impl Into<Option<&'a [i32]>>,
```
Comment (minghuaw) [inline src/random.rs:133]: This would accept `&[i32; N]` as args
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```

File: src/random.rs [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Generate normally distributed random numbers.
+///
+/// Generate an array of random numbers using the optional shape. The result
+/// will be of the given `T`. `T` must be a floating point type.
+///
+/// ```rust
```
Comment (minghuaw) [inline src/random.rs:116]: I think rust docs usually have the exampls at the bottom

File: src/random.rs [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+/// let key = mlx_rs::random::key(0);
+///
+/// // generate a single Float with normal distribution
```
Comment (minghuaw) [inline src/random.rs:119]:
```suggestion
/// // generate a single f32 with normal distribution
```

File: src/random.rs [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+/// let value = mlx_rs::random::normal::<f32>(None, None, None, &key).unwrap().item::<f32>();
+///
+/// // generate an array of Float with normal distribution in shape [10, 5]
```
Comment (minghuaw) [inline src/random.rs:122]:
```suggestion
/// // generate an array of f32 with normal distribution in shape [10, 5]
```

File: src/random.rs [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+/// let array = mlx_rs::random::normal::<f32>(&[10, 5][..], None, None, &key);
+/// ```
+///
+/// # Params
```
Comment (minghuaw) [inline src/random.rs:126]: I think there should be an empty line below # Params

File: src/random.rs [hunk 8]
```diff
@@ -0,0 +1,722 @@
...
+/// The matrix `covariance` must be positive semi-definite. The behavior is
+/// undefined if it is not.  The only supported output type is f32.
+///
+/// # Params
+/// - mean: array of shape `[..., n]`, the mean of the distribution.
+/// - covariance: array  of shape `[..., n, n]`, the covariance
+/// matrix of the distribution. The batch shape `...` must be
+/// broadcast-compatible with that of `mean`.
+/// - shape: The output shape must be
+/// broadcast-compatible with `mean.shape.dropLast()` and `covariance.shape.dropLast(2)`.
```
Comment (minghuaw) [inline src/random.rs:176]: What's the equivalent of `dropLast` in rust?

File: src/random.rs [hunk 9]
```diff
@@ -0,0 +1,722 @@
...
+///  - loc: mean of the distribution, default is `0.0`
+///  - scale: standard deviation of the distribution, default is `1.0`
+///  - key: PRNG key
+#[default_device]
+pub fn normal_device<'a, T: ArrayElement>(
+    shape: impl Into<Option<&'a [i32]>>,
+    loc: impl Into<Option<f32>>,
+    scale: impl Into<Option<f32>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:137]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 10]
```diff
@@ -0,0 +1,722 @@
...
+/// If empty, the result shape is determined by broadcasting the batch
+/// shapes of `mean` and `covariance`.
+/// - key: PRNG key
+#[default_device]
+pub fn multivariate_normal_device<'a, T: ArrayElement>(
+    mean: &Array,
+    covariance: &Array,
+    shape: impl Into<Option<&'a [i32]>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:186]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 11]
```diff
@@ -0,0 +1,722 @@
...
+/// scalars or arrays and must be roadcastable to `shape`.
+///
+/// ```rust
+/// use mlx_rs::{array, random};
+///
+/// let key = random::key(0);
+///
+/// // generate an array of Int values, one in the range 0 ..< 10
+/// // and one in the range 10 ..< 100
+/// let array = random::randint::<_, i32>(array![0, 10], array![10, 100], None, &key);
```
Comment (minghuaw) [inline src/random.rs:227]: Is there a better example? Maybe one in the range [0, 20), and another in the range [10, 100)? I think the current one seems confusing like someone may think the first argument specifies both lb and ub for the first rand int

File: src/random.rs [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+/// ```
+#[default_device]
+pub fn randint_device<'a, E: Into<Array>, T: ArrayElement>(
+    lower_bound: E,
+    upper_bound: E,
+    shape: impl Into<Option<&'a [i32]>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:235]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+///
+/// // generate an array of [3] Bool with the given p values
+/// let array = random::bernoulli(&array![0.1, 0.5, 0.8], None, &key);
+/// ```
+#[default_device]
+pub fn bernoulli_device<'a>(
+    p: impl Into<Option<&'a Array>>,
+    shape: impl Into<Option<&'a [i32]>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:291]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 14]
```diff
@@ -0,0 +1,722 @@
...
+/// // generate an array of shape [50, 2] of random Bool with p = 0.8
+/// let array = random::bernoulli(&p, &[50, 2][..], &key);
+///
+/// // generate an array of [3] Bool with the given p values
+/// let array = random::bernoulli(&array![0.1, 0.5, 0.8], None, &key);
+/// ```
+#[default_device]
+pub fn bernoulli_device<'a>(
+    p: impl Into<Option<&'a Array>>,
+    shape: impl Into<Option<&'a [i32]>>,
```
Comment (minghuaw) [inline src/random.rs:289]:
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```

File: src/random.rs [hunk 15]
```diff
@@ -0,0 +1,722 @@
...
+/// use mlx_rs::{array, Array, random};
+///
+/// let key = random::key(0);
+///
+/// // generate a single random Bool with p = 0.8
+/// let p: Array = 0.8.into();
+/// let value = random::bernoulli(&p, None, &key);
+///
+/// // generate an array of shape [50, 2] of random Bool with p = 0.8
+/// let array = random::bernoulli(&p, &[50, 2][..], &key);
```
Comment (minghuaw) [inline src/random.rs:281]:
```suggestion
/// let array = random::bernoulli(&p, &[50, 2], &key);
```

File: src/random.rs [hunk 16]
```diff
@@ -0,0 +1,722 @@
...
+/// will be of the given `T`. `T` must be a floating point type.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// // generate a single Float with normal distribution
+/// let value = mlx_rs::random::normal::<f32>(None, None, None, &key).unwrap().item::<f32>();
+///
+/// // generate an array of Float with normal distribution in shape [10, 5]
+/// let array = mlx_rs::random::normal::<f32>(&[10, 5][..], None, None, &key);
```
Comment (minghuaw) [inline src/random.rs:123]:
```suggestion
/// let array = mlx_rs::random::normal::<f32>(&[10, 5], None, None, &key);
```

File: src/random.rs [hunk 17]
```diff
@@ -0,0 +1,722 @@
...
+            )
+        };
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Generate values from a truncated normal distribution between `low` and `high`.
+///
+/// The values are sampled from the truncated normal distribution
+/// on the domain `(lower, upper)`. The bounds `lower` and `upper`
```
Comment (minghuaw) [inline src/random.rs:323]: Maybe change the args to `lower` and `upper` for consistency?

File: src/random.rs [hunk 18]
```diff
@@ -0,0 +1,722 @@
...
+///
+/// // generate an array of two Float values, one in the range 0 ..< 10
+/// // and one in the range 10 ..< 100
+/// let value = random::truncated_normal::<_, f32>(array![0, 10], array![10, 100], None, &key);
+/// ```
+#[default_device]
+pub fn truncated_normal_device<'a, E: Into<Array>, T: ArrayElement>(
+    lower_bound: E,
+    upper_bound: E,
+    shape: impl Into<Option<&'a [i32]>>,
```
Comment (minghuaw) [inline src/random.rs:339]:
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```

File: src/random.rs [hunk 19]
```diff
@@ -0,0 +1,722 @@
...
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:341]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 20]
```diff
@@ -0,0 +1,722 @@
...
+///
+/// // generate a single Float with Gumbel distribution
+/// let value = mlx_rs::random::gumbel::<f32>(None, &key).unwrap().item::<f32>();
+///
+/// // generate an array of Float with Gumbel distribution in shape [10, 5]
+/// let array = mlx_rs::random::gumbel::<f32>(&[10, 5][..], &key);
+/// ```
+#[default_device]
+pub fn gumbel_device<'a, T: ArrayElement>(
+    shape: impl Into<Option<&'a [i32]>>,
```
Comment (minghuaw) [inline src/random.rs:387]:
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```

File: src/random.rs [hunk 21]
```diff
@@ -0,0 +1,722 @@
...
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:389]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 22]
```diff
@@ -0,0 +1,722 @@
...
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// let logits = mlx_rs::Array::zeros::<u32>(&[5, 20]).unwrap();
+///
+/// // produces Array of u32 shape &[5]
+/// let result = mlx_rs::random::categorical(&logits, None, None, None, &key);
+/// ```
+///
+/// - Parameters:
```
Comment (minghuaw) [inline src/random.rs:430]: doc all params and move params above example?

File: src/random.rs [hunk 23]
```diff
@@ -0,0 +1,722 @@
...
+///     - logits: The *unnormalized* categorical distribution(s).
+#[default_device]
+pub fn categorical_device<'a>(
+    logits: &Array,
+    axis: impl Into<Option<i32>>,
+    shape: impl Into<Option<&'a [i32]>>,
```
Comment (minghuaw) [inline src/random.rs:436]:
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```

File: src/random.rs [hunk 24]
```diff
@@ -0,0 +1,722 @@
...
+    num_samples: impl Into<Option<i32>>,
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
```
Comment (minghuaw) [inline src/random.rs:439]:
```suggestion
    stream: impl AsRef<Stream>,
```

File: src/random.rs [hunk 25]
```diff
@@ -0,0 +1,722 @@
...
+            state.next()
+        },
+        |key| key.clone(),
+    );
+
+    let shape = shape.into();
+    let num_samples = num_samples.into();
+
+    // if we have both shape and num_samples, we need to error out
+    if shape.is_some() && num_samples.is_some() {
```
Comment (minghuaw) [inline src/random.rs:455]: I think this can be solved by introducing a new type?

```rust
pub enum ShapeOrCount {
    Shape(&[i32]),
    Count(i32)
}
```

Then the arg takes `shape_or_count: impl Into<Option<ShapeOrCount>>` (or `IntoOption<ShapeOrCount>`).

File: src/random.rs [hunk 26]
```diff
@@ -0,0 +1,722 @@
...
+                    axis,
+                    key.as_ptr(),
+                    stream.as_ptr(),
+                )
+            };
+            Ok(Array::from_ptr(c_array))
+        }
+    }
+}
+
```
Comment (minghuaw) [inline src/random.rs:503]: Is `random_bits` missing? There's `mlx_random_bits` in the c binding

**Review by minghuaw — COMMENTED** [review]
Left some comment for some minor issues. I think it's good to go once these are resolved

File: src/random.rs (hunk 2 above)
Comment (dcvz) [reply src/random.rs:80]: It is in python so I'll make it that way here too.

File: src/random.rs (hunk 26 above)
Comment (dcvz) [reply src/random.rs:503]: I saw that but I couldn't find usage of it anywhere in the swift or python libs

Comment (dcvz) [reply src/random.rs:503]: Seems like its used internally by the PRNG. I think I'd leave this out until someone needs it.

File: src/array/mod.rs [hunk 27]
```diff
@@ -443,21 +443,16 @@ impl Array {
     }
 }
 
-impl From<bool> for Array {
-    fn from(val: bool) -> Self {
-        Array::from_bool(val)
+impl<T: ArrayElement> From<T> for Array {
+    fn from(val: T) -> Array {
+        let slice = [val];
+        Array::from_slice(&slice, &[])
```
[the remark this answers is not in the document]
Comment (dcvz) [reply src/array/mod.rs:449]: Our tests would've failed right where it checks `ndim() == 0`? Being able to go from ArrayElement gives us some benefits.

**Review by minghuaw — APPROVED** [review]
LGTM!
